### PR TITLE
Morphy implementation draft

### LINKED ISSUE #19: Implement Morphy for English
"Morphy" is a very simple lemmatizer that is commonly included with English wordnet lookup. It should be included in this repository for completeness.

The definition of Morphy's behavior is here: https://wordnet.princeton.edu/documentation/morphy7wn

The NLTK has a Python implementation [here](https://github.com/nltk/nltk/blob/385fb3f22031ebfa87b5c5571f32008f967264fa/nltk/corpus/reader/wordnet.py#L1798-L1916), but I think we could follow the link above to recreate the algorithm from scratch. It would then be useful to look for differences with the NLTK. For instance, the NLTK has a `check_exceptions` parameter just checks if the form is in the list of exceptions (see Exception Lists in the Princeton link above), but we don't keep such lists.

### PR DESCRIPTION
**Pull request by tomaarsen**
Closes #19 

Hello!

This PR is a *draft* for the Morphy lemmatizer described in #19, meant for **discussion**. 
Everything here is up for debate, and nothing is set in stone. In particular, locations of functions and comments are not yet where they ought to be for release.

---

### Goals
The design document for morphy, as linked in #19, can be found [here](https://wordnet.princeton.edu/documentation/morphy7wn), and includes some ideas for an implementation. It also includes some goals, i.e. some functionality that a proper implementation should have. I'll go over them here (my case numbering):
* Morphy should function as a generator, yielding base forms until no more can be found.
* Morphy should handle:
  * Case 1: Hyphenations: `mother-in-law`, `son-of-a-gun`
  * Case 2: Collocations: `attorneys general`, `ask for it`, `customs duty`, `car pool`, `eat out`, `on them`
  * Case 3: Words ending in `-ful`: `boxesful`
  * Case 4: Standard words: `apples`, `ate`, `accepts`
  * Case 5: Words containing periods: `apples.`

Before I dive any deeper, this proposed implementation matches **all** of these goals.
It should be noted that `NLTK` only properly handles standard words, but does so considerably more effectively than this implementation. This is due to `NLTK`'s addition of an exception list. This is expanded upon in the Limitations section.

### Implementation
**Cases 1 through 3**
The implementation splits up these 5 cases, and handles them individually. The first three cases that are handled, collocations, hyphenations and words ending in `-ful`, rely on rules using regular expressions. Before being converted to regexes, these rules have readable formats, and are defined as such:
```python
HYPHENATED_RULES = expand_regex_rules({
    "sons-of-(PL)":        "son-of-a-(SING)",     # sons-of-guns      -> son-of-a-gun
    "(PL)-(PREP)-(ANY)":   "(SING)-(PREP)-(ANY)", # mothers-in-law    -> mother-in-law
    "(PREP)-(PL)":         "(PREP)-(SING)",       
    "(PL)-errant":         "(SING)-errant",       # knights-errant    -> knight-errant
    "(PL)-(PREP)":         "(SING)-(PREP)",       # passers-by        -> passer-by
    "(PL)-general":        "(SING)-general"
})

COLLOCATION_RULES = expand_regex_rules({
    "(PREP) them":         "(PREP) it",           # Only works for "on them" and "to them"
    "(ANY)star generals":  "(ANY)star general",
})

# Add all rules from hyphenated, but with the dash replaced with a space.
COLLOCATION_RULES.update({
    plural.replace("-", " "): lemma.replace("-", " ") for plural, lemma in HYPHENATED_RULES.items()
})

FUL_RULES = expand_regex_rules({
    "(PL)ful": "(SING)ful"
})
```
These rules are loosely based on Damian Conway's [Inflexion](https://metacpan.org/pod/Lingua::EN::Inflexion) library for Perl, and my thesis surrounding a port of that library. 
`(PL)` should be read as `plural`, and must correspond to a `(SING)` on the right-hand side. This is where morphy will recursively call morphy on the (supposedly) plural word, and place the lemmas of that call on the `(SING)` position.
`(PREP)` and `(ANY)` are simpler. They mean `preposition` and `anything` (i.e. non-empty) respectively, and must also correspond to that same tag used on the right-hand side of the rule. These values are not recursed with, and are preserved to the right-hand side.

These rules are kept readable so they can more easily be edited, but will be converted to a more useful format by the program itself. For example, the first rule in `HYPHENATED_RULES` will become:
```python
{
    re.compile('sons-of-(?P<rec0>.+)', re.IGNORECASE): 'son-of-a-{}'
}
```

These rules can now be tried. `try_rules` is called using a set of rules and a `word`, which is passed to `morphy` initially. This function will attempt to fullmatch the left-hand side pattern. If it manages to, it will call `morphy` with the subwords that need to be recursed on. We are left with a list of iterators, e.g. `[<generator object morphy at 0x000...>, ['out']]` where the generator object is the call to morphy with e.g. `"eats"`.

This creates a list of options, where each option is an iterator of potential lemmas that can be placed in their respective position on the right-hand side of the rule (e.g. in `son-of-a-{}`.) This list of options is passed to `expand_gen`, which will yield all possible combinations of options. e.g.:
```python
>>> list(expand_gen([1, 2, 3],
                    [4, 5],
                    [6, 7, 8]))
[
   [1, 4, 6], [2, 4, 6], [3, 4, 6], 
   [1, 5, 6], [2, 5, 6], [3, 5, 6], 
   [1, 4, 7], [2, 4, 7], [3, 4, 7], 
   [1, 5, 7], [2, 5, 7], [3, 5, 7], 
   [1, 4, 8], [2, 4, 8], [3, 4, 8], 
   [1, 5, 8], [2, 5, 8], [3, 5, 8]
]
>>> list(expand_gen([["asks", "ask"], ["for"], ["troubles", "trouble"]]))
[
    ['asks', 'for', 'troubles'], 
    ['ask', 'for', 'troubles'], 
    ['asks', 'for', 'trouble'], 
    ['ask', 'for', 'trouble']
]
```
This means that all options will be considered. These options will now be thrown into the right-hand side lemma format.
This is great for cases like `customs duty` where only one of the two words that can be modified *needs* to be modified to find the correct lemma: the one that is in Wordnet.

Then lastly, for each of these 3 cases (collocations, hyphenations, -ful) we check whether the lemma has not yet been yielded previously, and whether the lemma is in Wordnet or not.

**Case 1 and 2 extended**
For collocations and hyphenations, more is done: If none of the rules matched, but the input does contain multiple words (i.e. there is a space/dash in the already stripped input), then morphy will be called on each of these subwords. Each combination of outputs is tested to see if they are in Wordnet, and not a duplicate, like before, and yielded if so.

**Case 4**
After these three cases are handled, regular words have their turn. These are very simple. There are replacement rules with suffixes and replacements for each POS, and they are applied whenever the word matches the suffix. Whenever this happens, the lemma is tested like before.

**Case 5**
Then, inputs that have not matched anything else, that also contain a period, will recurse using a version of the input without any periods. 

Lastly, the input itself is also tested to see if it exists in Wordnet (and hasn't been yielded yet), and will be yielded if so.

---

### Notes
**Note 1**
The implementation will remove whitespace at the start and end of the input before processing. If `strip` is False, the default value, then this whitespace will be added back to any lemma that is yielded. As such, 
```python
>>> list(morphy('  apples  ', pos=NOUN))
['  apple  ']
>>> list(morphy('  apples  ', pos=NOUN, strip=True))
['apple']
```

**Note 2**
The implementation must have Python 3.7+ as it requires dict insertion order to be fixed, which (to my knowledge) was guaranteed starting from Python 3.7+, though was already often the case in Python 3.6, where it was an implementation detail.

**Note 3**
Due to the addition of `expand_gen`, one of the bugs listed in the design document does not exist in this implementation. The bug description: 
> Since many noun collocations contains prepositions, such as **line of products**, an algorithm similar to that used for verbs should be written for nouns. In the present scheme, if Morphy is passed **lines of products**, the search string becomes **line of product**, which is not in WordNet

Note that the *algorithm similar to that used for verbs* refers to the idea of inputs with the format `(verb) (preposition) (noun)` being handled accordingly.

**Note 4**
There are imperfections in the code. Some duplication remains still, commented code, etc. This is merely a draft after all.

---

### Limitations
**Limitation 1**
This implementation is severely disappointing due to the very reason that morphological generation is difficult: edge cases. The lack of any exception list, like `NLTK` has, and like the design document proposes, means that thousands of simple edge cases like `geese`, `oxen`, `feet`, etc. will not lemmatize properly. 
Even with the impressive performance for e.g. `'sons of bitches' -> ['son of a bitch']` or `'mothers-in-law' -> ['mother-in-law', 'mothers-in-law']` through this implementation's custom rules, this implementation will not be preferred over `NLTK`'s simpler implementation.

With that glaring flaw out of the way, let's discuss some smaller limitations:
**Limitation 2**
Capitalisation is not handled like what might be preferred yet. E.g. `morphy('Apples', NOUN)` will yield nothing, as the found lemma `Apple` is not in Wordnet. Only `apple` is. This can be resolved by forcing lowercase, but some Wordnet entries include uppercase characters, such as `Thompson submachine gun` and `Hanging Gardens of Babylon`.

**Limitation 3**
Currently, when recursing for subwords in collocations and hypenations, all subwords are the same POS as the combined word. This should be changed. `morphy("asking for troubles", NOUN)` yields nothing, despite `ask for trouble` being in Wordnet, as `morphy("troubles", wn.VERB)` yields nothing. This is mentioned in the design document as well: that collocations often have the form `(verb) (preposition) (noun)`. One way to change this is to no longer force a call to `morphy` to include a POS. When there is no POS given, the rules for all POS' would be tested, and it would be tested whether the lemma exists in Wordnet for *any* POS.
Alternatively, an additional parameter is added to `morphy` that allows non-Wordnet words to be yielded. 

---

### Rough tests
I've not set up a complete testing suite yet, but have created a small file containing some words to be passed to morphy: 
```python
from wn.morphy import morphy
from wn.constants import NOUN, VERB, ADJ

def quick_morphy_test(word: str, pos: str = NOUN, strip: bool = False) -> List[str]:
    gen = morphy(word, pos, strip)
    print(f"{word!r} -> {list(gen)}")

if __name__ == "__main__":
    quick_morphy_test("Apples")
    quick_morphy_test("  apples  ")
    quick_morphy_test("  apples  ", strip=True)
    quick_morphy_test("cars pools")
    quick_morphy_test("sons of bitches")
    quick_morphy_test("attorneys general")
    quick_morphy_test("asks for it", pos=VERB)
    quick_morphy_test("customs duties")
    quick_morphy_test("asking for it", pos=VERB)
    quick_morphy_test("asking for troubles", pos=VERB) # This fails
    quick_morphy_test("asking for trouble", pos=VERB)  # This works
    quick_morphy_test("apples.")
    quick_morphy_test("geese")
    quick_morphy_test("oxen")
    quick_morphy_test("feet")
    quick_morphy_test("mothers-in-laws")
    pass
```
This currently outputs:
```python
'Apples' -> []
'  apples  ' -> ['  apple  ']
'  apples  ' -> ['apple']
'cars pools' -> ['car pool']
'sons of bitches' -> ['son of a bitch']
'attorneys general' -> ['attorney general']
'asks for it' -> ['ask for it']
'customs duties' -> ['customs duty']
'asking for it' -> ['ask for it']
'asking for troubles' -> []
'asking for trouble' -> ['ask for trouble']
'apples.' -> ['apple']
'geese' -> ['geese']
'oxen' -> ['oxen']
'feet' -> ['feet']
'mothers-in-laws' -> ['mother-in-law', 'mothers-in-law']
```
Tests that have actual purpose to try to guarantee the accuracy of `morphy` are yet to be written.

---

I would very much like to see the exception list be added to `wn` - to be able to conclusively state that this module has a better implementation for morphy than `NLTK`. 

---

I'm looking for as much feedback as I can get. This has been fun thus far, and is right up my alley.

- Tom Aarsen

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -102,7 +102,7 @@ celerybeat.pid
 *.sage.py
 
 # Environments
-.env
+.env*
```
Comment (goodmami) [inline .gitignore:105]: This change is unrelated to the point of the PR

File: wn/morphy.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+}
+
+PREPOSITIONS = ["about", "above", "across", "after", "among", "around", "athwart", 
+                "at", "before", "behind", "below", "beneath", "beside", "besides", 
+                "between", "betwixt", "beyond", "but", "by", "during", "except", 
+                "for", "from", "into", "in", "near", "off", "of", "onto", "on", 
+                "out", "over", "since", "till", "to", "under", "until", "unto", 
+                "upon", "with"]
+
+def expand_regex_rules(rules: dict) -> dict:
```
Comment (goodmami) [inline wn/morphy.py:47]: Functions that aren't meant to be called directly by users or outside of this module should be prefixed with `_`

File: wn/morphy.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        except StopIteration:
+            raise IndexError
+
+def expand_gen(options: Iterable[Iterable[str]]) -> Iterable[Iterable[str]]:
+    for sublemmas in _expand_gen([LazyList(option) for option in options]):
+        yield sublemmas
+"""
+
+def morphy(word: str, pos: str, strip:bool = False) -> Generator[str, None, None]:
+    """
```
Comment (goodmami) [inline wn/morphy.py:147]: Public functions should have a 1-line summary at the top of the docstring.

File: wn/morphy.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    def reapply_whitespace(lemma: str, match: re.Match, strip: bool) -> str:
+        if strip:
+            return lemma
+        return match.group("before") + lemma + match.group("after")
+
+    yielded_lemmas = []
+
+    def in_wordnet(lemma: str, pos: str) -> bool:
+        try:
+            next(find_entries(form=lemma, pos=pos))
```
Comment (goodmami) [inline wn/morphy.py:224]: The call to `find_entries()` should specify the `lexicon_rowids`, which means they need to come from somewhere.

File: wn/morphy.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+            # if not yielded_previously(lemma):# and in_wordnet(lemma, pos):
+            yielded_lemmas.append(lemma)
+            yield reapply_whitespace(lemma, ws_match, strip)
+
+    # Step 7: Yield the input word if it's already in Wordnet
+    if not yielded_previously(word) and in_wordnet(word, pos):
+        yielded_lemmas.append(word)
+        yield reapply_whitespace(word, ws_match, strip)
+
+__all__ = [
```
Comment (goodmami) [inline wn/morphy.py:304]: Conventionally, `__all__` goes at the top of the module.

**Review by goodmami — COMMENTED** [review]
Thanks! I had no idea you were working on this.

In general the code looks nice and clean. I did a review, but it was just the low-hanging fruit and some style suggestions. I haven't yet looked deeply at the code or tested it.

A couple of questions for now:
* Does this exactly implement the behavior of morphy as defined at the Princeton link in #19? It's ok to improve upon it, but those improvements should be optional.
* Are there parts of the algorithm that don't require DB lookups? If so, it would nice to have some unit tests for the code.

**Comment by arademaker** [comment]
What can we push to data to improve multilingual support?

**Comment by goodmami** [comment]
> What can we push to data to improve multilingual support? 

@arademaker see #8. Morphy is an English-only method.

**Comment by CubieDev** [comment]
@goodmami 
Some personal things have popped up recently, that have taken priority. I'm forced to put this implementation on the backburner for now, but I hope to return to it shortly.

**Comment by goodmami** [comment]
@CubieDev There's no big rush. Take your time.

**Comment by goodmami** [comment]
@CubieDev just checking in. Do you think you'll have time to work on this in the near future? If so I'll take a closer look at the submission and recommend some other things. For now, here are a few more thoughts in addition to my comments above:

* Regarding Python 3.7; this library supports Python 3.6, which still has about a year of life left. Until 3.6 support is dropped, we should try to accommodate 3.6 in some way. You're correct that ordered dictionaries existed in 3.6 but that it was an implementation detail. That is, it might not be true for, e.g., Pypy or some other implementation.
* If possible, I'd like to avoid importing the morphy module on `import wn`. Maybe only when an English wordnet is loaded?
* Beyond the `wn.morphy` module, do you have thoughts about how to incorporate it into user queries (e.g., `wn.synsets("dogs", normalize=morphy)` or something)?

**Comment by CubieDev** [comment]
@goodmami I have recently finished my thesis, and am about to finish up my work for the quarter. If all goes well, some semblance of free time should be available to me again in a week or two. By then, I'll be able to address your comments.
I hope the rest of the development work has been going smoothly in my absence.

**Comment by goodmami** [comment]
Good, glad to hear, and congrats on finishing your thesis!

When you're ready, let's start by addressing my comments above and then we can drill down further if necessary.

**Comment by CubieDev** [comment]
@goodmami 
I'm afraid I'm being drowned in projects, and am going to struggle to get enough spare time to work on this soon. I wish I had better news for you, as I love working on open sourced projects like these, but that is the reality for now. 

I'd love to revisit this in the future, but I can't give you a guarantee that I'll get around to it shortly. I figured I should let you know. 

- Tom Aarsen

**Comment by goodmami** [comment]
@CubieDev no worries. I'll leave this open for now because there's good work in there. Hopefully I'll get a chance to find a way to fit it in. Thanks for the update!

**Comment by goodmami** [comment]
As an update, I just looked at how I could merge this in, but the necessary edits would have been substantial, so I started a new implementation in the 'morphy' branch. It's coming from a different angle: instead of trying to handle many morphological patterns, it starts simple (similar to the NLTK implementation) but focuses on how it can integrate with Wn's API (see #8). Currently it only does single words (the "rules of detachment" plus some added rules) and those listed as alternate forms in the wordnet file (which I've found is not the same as the exception lists; e.g. the English WordNet 2020 doesn't have "axes" -> "axis"). Later, I hope to pull features out of this PR to augment the morphological analysis of the simpler algorithm. The problem with this is I won't be able to retain the author of the original commits, so we might have to settle for attribution in a commit message.

**Comment by tomaarsen** [comment]
Hello! (As you may have noticed, I have renamed my account from CubieDev to tomaarsen)
I figured that my draft implementation would quickly stop working effectively as this module continues to quickly expand. Furthermore I support the idea of starting simple and then opting to expand afterwards. 
Once you're satisfied with your simple morphy implementation, feel free to close to PR draft and let me know. I might be able to apply some of my practical morphological analysis and generation knowledge that I acquired through my thesis project: https://github.com/tomaarsen/Inflexion
This project deals with simple detection of collocations, allowing them to be correctly inflected, e.g. `mothers in law` -> `mother in law`.

And regarding the commit ownership - don't worry about it. Attribution in a commit message works just fine, and if you draw enough inspiration from my work, feel free to plop my name in a contributors file. 

I'll definitely stick around and follow the progression of this work as it develops, and see if I can pop in for some help at times.

- Tom Aarsen

**Comment by goodmami** [comment]
Ok, I've merged in the new `wn.morphy` module, so I think it's time to close this PR. The new module is currently very simple (about the same as the NLTK but without recursive rule application), but I adapted the algorithm for how queries work in Wn. For example, the basic behavior returns any successful rule applications, including non-words, because subsequent queries against the database will perform the filtering that is normally done within Morphy, but users can optionally instantiate a Morphy object with a wordnet to get more proper lemmatization behavior.

Overall, for the current iteration I was more concerned with the API design (see the docs on [lemmatization](https://wn.readthedocs.io/en/latest/guides/lemmatization.html) and [wn.morphy](https://wn.readthedocs.io/en/latest/api/wn.morphy.html)) and performance and less so the sophistication of the lemmatization. For next steps I'd like to refer to your work here and to the original C implementation of Morphy to get collocations, -ful words, etc., and hopefully the ability to choose a compatibility mode (e.g., NLTK or PWN compatibility), for which I've laid some groundwork already. And, of course, I'm happy to receive future PRs for these things!

BTW I don't have a contributors file so I added your name in the [release notes](https://github.com/goodmami/wn/releases/tag/v0.7.0).

Thanks again!
